### PR TITLE
Refactoring to use profiles.tuned.openshift.io

### PR DESCRIPTION
**Pull request by jmencak**
This is a largish refactoring of the NTO's operand, openshift-tuned.
Changes:
  - Do not pull any node/pod information from the API server.  Use
    a shared informer for profiles.tuned.openshift.io objects which
    carry a precalculated tuned profile for the node.
  - Switch to klog, glog seems unmaintained and vendored-in libraries
    already depend on klog.
  - Added tuna and tuned-profiles-cpu-partitioning profiles
  - Redesign of the retry loop
  - Makefile/.gitignore cleanup
  - Added vendor dependencies

### CODE REVIEW

**Comment by jmencak** [comment]
/hold
This is a 4.4 branch code, do not merge yet.
Needs to merge together with https://github.com/openshift/cluster-node-tuning-operator/pull/96

File: cmd/openshift-tuned.go [hunk 1]
```diff
@@ -652,128 +411,103 @@ func timedTunedReloader(tuned *tunedState) (err error) {
...
+	si := cache.NewSharedInformer(profileLW, &tunedv1.Profile{}, 0)
+	si.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			profile, err := getTunedProfile(obj)
+			if err != nil {
+				klog.Errorf("%s", err.Error())
+				return
+			}
+			klog.V(1).Infof("profile %q added, tuned profile requested: %s", profile.ObjectMeta.Name, profile.Spec.Config.TunedProfile)
+			disableSystemTuned() // Whe moving this call elsewhere, remember it is undesirable to disable system tuned
```
Comment (sjug) [inline cmd/openshift-tuned.go]: Small typo should probably be "When"?

File: cmd/openshift-tuned.go [hunk 2]
```diff
@@ -652,128 +411,103 @@ func timedTunedReloader(tuned *tunedState) (err error) {
...
+	stop := make(chan struct{})
+	defer close(stop)
+	go si.Run(stop)
 
-	// Create a ticker to dump node/pod labels, extract new profiles and possibly reload tuned;
-	// this also rate-limits reloads to a maximum of labelDumpInterval reloads/s
-	tickerReload := time.NewTicker(time.Second * time.Duration(labelDumpInterval))
+	// Create a ticker to extract new profiles and possibly reload tuned;
+	// this also rate-limits reloads to a maximum of profileExtractInterval reloads/s
+	tickerReload := time.NewTicker(time.Second * time.Duration(profileExtractInterval))
```
Comment (sjug) [inline cmd/openshift-tuned.go:489]: Why did we want to shorten the ticker time to every 1s?

File: cmd/openshift-tuned.go [hunk 3]
```diff
@@ -813,61 +547,51 @@ func (resyncPeriod *resync) changeWatcher() (err error) {
...
-	resyncPeriod.min = resyncPeriod.now
+	const (
+		errsMax        = 5  // the maximum number of consecutive errors within errsMaxWithinSeconds
+		sleepRetryInit = 10 // the initial retry period [s]
+	)
+	var (
+		errs       int
+		sleepRetry int64 = sleepRetryInit
+		// sum of the series: S_n = x(1)*(q^n-1)/(q-1) + add 60s for each changeWatcher() call
+		errsMaxWithinSeconds int64 = (sleepRetry*int64(math.Pow(2, errsMax)) - sleepRetry) + errsMax*60
```
Comment (sjug) [inline cmd/openshift-tuned.go:589]: What is the variable `errsMaxWithinSeconds`?

File: cmd/openshift-tuned.go [hunk 4]
```diff
@@ -652,128 +411,103 @@ func timedTunedReloader(tuned *tunedState) (err error) {
...
-
-func (resyncPeriod *resync) changeWatcher() (err error) {
+func changeWatcher() (err error) {
 	var (
 		tuned tunedState
-		wPod  watch.Interface
 		lStop bool
 	)
 
+	namespace := "openshift-cluster-node-tuning-operator"
```
Comment (sjug) [inline cmd/openshift-tuned.go]: This seems constant

File: cmd/openshift-tuned.go [hunk 5]
```diff
@@ -652,128 +411,103 @@ func timedTunedReloader(tuned *tunedState) (err error) {
...
 	nodeName := flag.Args()[0]
+	fieldSelector := fields.SelectorFromSet(fields.Set{"metadata.name": nodeName})
```
Comment (sjug) [inline cmd/openshift-tuned.go]: You can toss the `nodeName` and `fieldSelector` in the `var` block as well for consistency

File: cmd/openshift-tuned.go [hunk 6]
```diff
@@ -438,207 +355,49 @@ func getActiveProfile() (string, error) {
...
-	for event := range w.ResultChan() {
-		f(event, tuned)
-	}
-}
-
 func timedTunedReloader(tuned *tunedState) (err error) {
 	var (
-		reload        bool
-		labelsChanged bool
+		reload bool
```
Comment (sjug) [inline cmd/openshift-tuned.go]: Probably don't need a multi line var decl here.

File: cmd/openshift-tuned.go [hunk 7]
```diff
@@ -198,59 +161,84 @@ func getConfig() (*rest.Config, error) {
...
+	return nil, fmt.Errorf("could not locate a kubeconfig")
 }
 
-func getJitter(period int64, factor float64) int64 {
-	return rand.Int63n(int64(float64(period)*factor+1)) - int64(float64(period)*factor/2)
+func disableSystemTuned() {
+	klog.V(1).Infof("disabling system tuned...")
+	cmd := exec.Command("/usr/bin/systemctl", "disable", "tuned", "--now")
+	var stdout bytes.Buffer
+	var stderr bytes.Buffer
```
Comment (sjug) [inline cmd/openshift-tuned.go]: Can combine to a single line declaration

File: cmd/openshift-tuned.go [hunk 8]
```diff
@@ -114,7 +95,19 @@ func (a *arrayFlags) Set(value string) error {
 	return nil
 }
 
+func newUnixListener(addr string) (net.Listener, error) {
```
Comment (sjug) [inline cmd/openshift-tuned.go]: Any reason for this function move?

**Review by sjug — COMMENTED** [review]
Some thoughts

File: cmd/openshift-tuned.go (hunk 1 above)
Comment (jmencak) [reply cmd/openshift-tuned.go]: Yes, thank you!

File: cmd/openshift-tuned.go (hunk 2 above)
Comment (jmencak) [reply cmd/openshift-tuned.go:489]: Because I believe we can now.  Faster change propagation.

File: cmd/openshift-tuned.go (hunk 3 above)
Comment (jmencak) [reply cmd/openshift-tuned.go:589]: It is a duration in seconds.  In particular the sum of the geometric series of `errsMax` sleepRetry(s) + some extra time for each retry.  `sleepRetry` in increased 2x for every failure.  With the current code (unlike prior to the rearchitecture), I have not seen *any* retries, but I'd say it still make some sense to have a retry function of some sort.

**Comment by zvonkok** [comment]
Looking good for me, this rework makes the operand as dumb as possible, the new informer and new design will minimize API access and resolve the issue with slow propagation of tuned profiles.

**Pull request merged**